File: BackEnd/Api/Domain/Entities/PrivacyPolicy.cs
namespace Api.Domain.Entities;

using Api.Domain.ValueObjects;
using Api.Application.DataTransfer;
using Api.Infrastructure.Utilities;

public class PrivacyPolicy : IDTOAble<PrivacyPolicyDto>
{
	public Guid Id { get; private set; }
	public string Content { get; private set; }
	public DateTime UpdatedOn { get; private set; }
	public bool Active { get; private set; } = true; // Soft delete

	protected PrivacyPolicy() { } // EF Core

	internal PrivacyPolicy(Guid id, string content, DateTime updatedOn, bool active)
	{
		Id = id;
		Content = content ?? throw new ArgumentNullException(nameof(content));
		UpdatedOn = updatedOn;
		Active = active;
	}

	public void Deactivate()
	{
		Active = false;
	}

	public PrivacyPolicy Update(string content)
	{
		UpdateContent(content);
		UpdateUpdatedOn(DateTime.UtcNow);
		return this;
	}

	private void UpdateContent(string content) { Content = content; }
	private void UpdateUpdatedOn(DateTime updatedOn) { UpdatedOn = updatedOn; }

	public PrivacyPolicyDto ToDTO()
	{
		return new PrivacyPolicyDto
		{
			Id = Id,
			Content = Content,
			UpdatedOn = UpdatedOn,
			Active = Active
		};
	}
}

File: FrontEnd/src/model/PrivacyPolicy.ts
import type PrivacyPolicyDto from "./dto/PrivacyPolicyDto";

export class PrivacyPolicy {

    public content: string;
    public updatedOn: Date;
    public active: boolean;

    constructor(
        params: {
            content: string;
            updatedOn: Date;
            active: boolean;
        }
    ) {
        if (!params.content) throw new Error('Content cannot be null or empty.');

        this.content = params.content;
        this.updatedOn = params.updatedOn;
        this.active = params.active;
    }

    isActive(): boolean {
        return this.active;
    }

    updateContent(content: string): void {
        if (!content) throw new Error('Content cannot be null or empty.');
        this.content = content;
    }

    deactivate(): void {
        this.active = false;
    }

    toDto(): PrivacyPolicyDto {
        return {
            content: this.content,
        };
    }
}